src/aspire/ctf/ctf_estimator.py

```diff
@@ -0,0 +1,757 @@
+            np.arange(N, dtype=self.dtype), np.arange(N, dtype=self.dtype)
+        )
+
+        denom = np.pi * (k - el)
+        denom = denom + np.eye(N, dtype=self.dtype)
+        phi_R = np.divide(np.sin(np.pi * R * (k - el)), denom)
+        np.fill_diagonal(phi_R, 1)  # absolute difference from Matlab 10^-18
+
+        data_tapers, _ = eigs(phi_R, L)
+
+        return data_tapers
+
+    def ctf_estimate_psd(self, blocks, tapers_1d, num_1d_tapers):
+        """
+        Estimate the power spectrum of the micrograph using the multi-taper method
+
+        :param blocks: 3-D numpy array containing windows extracted from the micrograph in the ctf_preprocess function.
+        :param tapers_1d: numpy array of data tapers.
+        :param num_1d_tapers: number of 1-D data tapers
+        :return: Numpy array of estimated power spectrum.
+        """
+
+        tapers_1d = tapers_1d.astype(complex_type(self.dtype), copy=False)
+
+        blocks_mt_pre_fft = np.zeros(
+            blocks[0, :, :].shape, dtype=complex_type(self.dtype)
+        )
+
+        blocks_mt_post_fft = np.zeros(
+            blocks_mt_pre_fft.shape, dtype=complex_type(self.dtype)
+        )
+
+        blocks_mt = np.zeros(blocks_mt_pre_fft.shape, dtype=self.dtype)
+
+        blocks_tapered = np.zeros(blocks[0, :, :].shape, dtype=complex_type(self.dtype))
+
+        taper_2d = np.zeros(
+            (blocks.shape[1], blocks.shape[2]), dtype=complex_type(self.dtype)
+        )
+
+        for x in range(num_1d_tapers ** 2):
+            np.matmul(
+                np.reshape(tapers_1d[:, x // num_1d_tapers], (tapers_1d.shape[0], 1)),
+                np.reshape(tapers_1d[:, x % num_1d_tapers], (1, tapers_1d.shape[0])),
+                out=taper_2d,
+            )
+            for m in range(blocks.shape[0]):
+                np.multiply(blocks[m, :, :], taper_2d, out=blocks_tapered)
+                blocks_mt_post_fft = fft.fftn(blocks_tapered, axes=(-2, -1))
+                blocks_mt = blocks_mt + abs2(blocks_mt_post_fft)
+
+        blocks_mt = blocks_mt / (blocks.shape[0])
+        blocks_mt = blocks_mt / (blocks.shape[0])
+        blocks_mt = blocks_mt / tapers_1d.shape[0] ** 2.0
+
+        thon_rings = np.fft.fftshift(
+            blocks_mt
+        )  # max difference 10^-13, max relative difference 10^-14
+
+        return Image(thon_rings)
+
+    def ctf_elliptical_average(self, ffbbasis, thon_rings, k):
+        """
+        Computes radial/elliptical average of the power spectrum
+
+        :param ffbbasis: FFBBasis instance.
+        :param thon_rings: Power spectrum.
+        :param k: 0 for radial averaging and 2 for elliptical averaging.
+        :return: PSD and noise as 2-tuple of numpy arrays.
+        """
+
+        # RCOPT, come back and change the indices for this method
+        coeffs_s = ffbbasis.evaluate_t(thon_rings).T
+        coeffs_n = coeffs_s.copy()
+
+        coeffs_s[np.argwhere(ffbbasis._indices["ells"] == 1)] = 0
```

Comment:
Maybe @ayeltg can help answer a couple of the implementation conversations (this one and the linprog setup).  I'd just be guessing/interpreting...

Reply:
The ffb basis object was defined as: ffbbasis = FFBBasis2D((psd_size, psd_size), 2, dtype=dtype), so any expansion over this basis will be limited to an expansion over the basis functions with ell=0, ell=1, ell=2. 
Also, we know that, due to the mathematical model of the CTF, the expansion over basis functions with ell=1 should be zero, that is, if we expand over these functions we are guaranteed to see noise, and this is what coeffs_n does.

Reply:
Ok so this information should be provided in the code. Whoever reads this may not know that `ell_max = 2` in the basis decomposition. We also don't check that `ell_max = 2`, so this could cause bugs. Might be safer to do something like
```
if circular:
	coeffs_s[ffbbasis._indices["ells"] > 0] = 0
else:
	coeffs_s[ffbbasis._indices["ells"] > 2] = 0

	# Coefficients with ell = 1 do not occur in the CTF (see eq. (XXX) in paper)
	coeffs_s[ffbbasis._indices["ells"] == 1] = 0
```

Now I'm still having trouble making sense of `coeffs_n`. If `circular = True`, we have that `noise = thon_rings` but if `circular = False`, it's equal to the `ell = 1` coefficients. Shouldn't it be the `ell = 1` and `ell = 2` coefficients in the `circular = True` case then?

Reply:
Created #422 so this discussion can continue later.